### ISSUE TITLE
Memory leak (not released) when calling Seq2SeqTrainer for fine-tuning

### ISSUE THREAD

**Issue by xyx361100238** [huggingface/transformers#27992]
### System Info

- `transformers` version: 4.34.0
- Platform: Linux-5.4.0-167-generic-x86_64-with-glibc2.31
- Python version: 3.9.18
- Huggingface_hub version: 0.17.3
- Safetensors version: 0.4.0
- Accelerate version: 0.23.0
- Accelerate config:    not found
- PyTorch version (GPU?): 1.13.1+cu116 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

@sanchit-gandhi  @muellerzr @pacman100

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

I was use project finetune.py from [Whisper-Finetune](https://github.com/yeyupiaoling/Whisper-Finetune)  to finetune whisper large on my own datasets，which code use transformers  library. If I use the validation process during fine-tuning, it will lead to an increase in system memory, and there is a chance that it will not be released after validation. Over time, this can lead to Out Of Memory (OOM) and cause a crash. Is this a bug in the tool, or do I need to make some settings?

### Expected behavior

After each verification is completed, normalize the memory to maintain it at a stable level.

**Comment by sanchit-gandhi** [comment]
Hey @xyx361100238! Do you have a script I can use to reproduce the error and identify the memory increase? It's maybe worth trying adding the flag for full fp16 eval as well to avoid upcasting the weights to fp32 for inference:

```
--fp16_full_eval
```

_c.f._ https://huggingface.co/docs/transformers/v4.36.1/en/main_classes/trainer#transformers.TrainingArguments.fp16_full_eval
